pkg/router/crl/crl.go

```diff
@@ -0,0 +1,211 @@
+				log.Info("certificate revocation list has expired", "subject key identifier", subjectKeyId, "next update", crl.TBSCertList.NextUpdate.Format(time.RFC3339))
+			} else {
+				subjectKeyIds = append(subjectKeyIds, subjectKeyId)
+				if (nextCRLUpdate.IsZero() || crl.TBSCertList.NextUpdate.Before(nextCRLUpdate)) && crl.TBSCertList.NextUpdate.After(now) {
+					nextCRLUpdate = crl.TBSCertList.NextUpdate
+				}
+				continue
+			}
+		}
+		log.Info("retrieving certificate revocation list", "subject key identifier", subjectKeyId)
+		if crl, err := getCRL(cert.CRLDistributionPoints, now); err != nil {
+			// Creating or updating the configmap with incomplete data would compromise security by potentially
+			// permitting revoked certificates.
+			return fmt.Errorf("failed to get certificate revocation list for certificate key %s: %w", subjectKeyId, err)
+		} else {
+			crls[subjectKeyId] = crl
+			subjectKeyIds = append(subjectKeyIds, subjectKeyId)
+			log.Info("new certificate revocation list", "subject key identifier", subjectKeyId, "next update", crl.TBSCertList.NextUpdate.Format(time.RFC3339))
+			if (nextCRLUpdate.IsZero() || crl.TBSCertList.NextUpdate.Before(nextCRLUpdate)) && crl.TBSCertList.NextUpdate.After(now) {
+				nextCRLUpdate = crl.TBSCertList.NextUpdate
+			}
+			updated = true
+		}
+	}
+
+	if len(subjectKeyIds) == 0 {
+		if !crlFileExists {
+			// Make sure the crl file exists, even if it's empty
+			return os.WriteFile(crlFilename, []byte(""), crlFilePermissions)
+		} else {
+			return nil
+		}
+	}
+
+	// If any CRLs changed, encode the CRLs and write to `crlFilename`
+	if updated {
+		buf := &bytes.Buffer{}
+		for _, subjectKeyId := range subjectKeyIds {
+			asn1Data, err := asn1.Marshal(*crls[subjectKeyId])
+			if err != nil {
+				return fmt.Errorf("failed to encode ASN.1 for CRL for certificate key %s: %w", subjectKeyId, err)
+			}
+			block := &pem.Block{
+				Type:  "X509 CRL",
+				Bytes: asn1Data,
+			}
+			if err := pem.Encode(buf, block); err != nil {
+				return fmt.Errorf("failed to encode PEM for CRL for certificate key %s: %w", subjectKeyId, err)
+			}
+		}
+		crlData := buf.String()
+
+		return os.WriteFile(crlFilename, []byte(crlData), crlFilePermissions)
+	}
+
+	return nil
+}
+
+// getCRL gets a certificate revocation list using the provided distribution points and returns the certificate list.
+func getCRL(distributionPoints []string, now time.Time) (*pkix.CertificateList, error) {
+	var errs []error
+	for _, distributionPoint := range distributionPoints {
+		// The distribution point is typically a URL with the "http" scheme.  "https" is generally not used because the
+		// certificate list is signed, and because using TLS to get the certificate list could introduce a circular
+		// dependency (cannot use TLS without the revocation list, and cannot get the revocation list without using
+		// TLS).
+		//
+		// TODO Support ldap.
+		switch {
+		case strings.HasPrefix(distributionPoint, "http:"):
```

Comment:
Maybe use `url.Parse(s)` and check the scheme. The string may begin with "http:" but the remainder could be bogus.

Reply:
I haven't made this change yet, but it does look like `url.Parse` would be a good choice here

Reply:
`getHTTPCRL` uses `http.Get`, and I expect `http.Get` will return an error if the URL is invalid, in which case using `url.Parse` here just for the extra error checking would be superfluous.  Really all we care about is whether the scheme is "http:" as opposed to "ldap:", so it makes sense to me to keep it simple.